src/falcon/signature.rs

```diff
@@ -0,0 +1,113 @@
+use super::{Polynomial, LOG_N, MODULUS, N, PK_LEN, SIG_L2_BOUND, SIG_LEN};
+
+// FALCON SIGNATURE
+// ================================================================================================
+
+/// A signature of a message
+///
+/// The signature is a pair of polynomials (s1, s2)
+/// where s1 = c - s2 * h
+/// and h is the public key.
+///
+/// The signature  verifies if and only if:
+/// 1. s1 = c - s2 * h
+/// 2. |s1|^2 + |s2|^2 <= SIG_L2_BOUND
+///
+/// where |.| is the norm.
+///
+/// `Signature` also includes the (raw) public key which is serialized as:
+/// 1. 1 byte representing the log of N i.e. 512.
+/// 2. 896 bytes for the public key. This is decoded into the `h` polynomial above.
+///
+/// The actual signature is serialized as:
+/// 1. 40 bytes for the nonce.
+/// 2. 626 bytes for the actual signature. This is decoded into the `s2` polynomial above.
+pub struct Signature {
+    sig: [u8; SIG_LEN],
+    pk: [u8; PK_LEN],
+}
+
+impl Signature {
+    /// Returns a new [Signature] instantiated from the provided signature and expanded public key.
+    pub fn new(sig: [u8; SIG_LEN], pk: [u8; PK_LEN]) -> Self {
+        Self { sig, pk }
+    }
+
+    /// Unpack the signature into a polynomial in Z_q[x]/(phi)
+    pub fn unpack(&self) -> [u16; N] {
+        let res = sig_decode(self.sig[41..].as_ref());
+        res
+    }
+
+    /// Return the nonce component of the signature
+    pub fn nonce(&self) -> &[u8] {
+        self.sig[1..41].as_ref()
+    }
+
+    /// Return the public key component of the signature
+    pub fn pk(&self) -> Polynomial {
+        assert!(self.pk[0] == LOG_N as u8);
+        Polynomial::from_pubkey(&self.pk[1..])
+    }
+
+    // Verify the signature
+    pub fn verify(&self, message: &[u8]) -> bool {
+        let h: Polynomial = self.pk();
+        let s2: Polynomial = self.into();
+        let c = Polynomial::from_hash_of_message(message, self.sig[1..41].as_ref());
+
+        let s1 = c - s2 * h;
+
+        let sq_norm = s1.sq_norm() + s2.sq_norm();
+        sq_norm <= SIG_L2_BOUND
+    }
```

Comment:
We should probably pass in the public key commitment here and verify that expanded public key hashes to it. I know we do this check in `PublicKey::verify()` but since this is also a part of public interface, we should probably move this check here.

Reply:
Makes sense!